huntsman/observatory.py

```diff
@@ -472,43 +355,217 @@ def _create_autoguider(self):
+                    time.sleep(30)
+            else:
+                if all_too_bright:
+                    self.logger.debug('All exposures are too bright. Waiting 30 seconds...')
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _autoflat_mean_counts(self, filename, bias, min_counts=1):
+        """ Read the data and calculate a clipped-mean count rate.
 
-        self.logger.debug("Flat-field observation: {}".format(flat_obs))
+        Args:
+            filename (str): The filename containing the data.
+            bias (float): The bias level to subtract from the image.
+            min_counts (float): The minimum count rate returned by this funtion.
+        """
+        try:
+            data = fits.getdata(filename)
+        except FileNotFoundError:
+            data = fits.getdata(filename + '.fz')
+
+        # Calculate average counts per pixel
+        mean_counts, _, _ = stats.sigma_clipped_stats(data-bias)
+        mean_counts = max(min_counts, mean_counts - bias)
+        return mean_counts
+
+    def _autoflat_next_exptime(self, previous_exptime, elapsed_time, target_counts, mean_counts):
+        """Calculate the next exposure time for the flat fields, accounting
+        for changes in sky brightness."""
+        exptime = previous_exptime * (target_counts / mean_counts)
+        sky_factor = 2.0 ** (elapsed_time / 180.0)
+        if self.past_midnight:
+            exptime = exptime / sky_factor
+        else:
+            exptime = exptime * sky_factor
+        return round(exptime.to_value(u.second)) * u.second
+
+    def _take_flat_observation(self, exptimes, observation, fits_headers=None, dark=False):
+        """
+        Slew to flat field, take exposures and wait for them to complete.
+        Returns a list of camera events for each camera.
+
+        args:
+            exptimes: dict of camera_name: list of exposure times.
+            observation: Flat field Observation object.
+        """
+        imtype = 'dark' if dark else 'flat'
+        if fits_headers is None:
+            fits_headers = self.get_standard_headers(observation=observation)
+
+        # Slew to field
+        self.logger.debug(f'Slewing to flat-field coords: {observation.field}.')
+        self.mount.set_target_coordinates(observation.field)
+        self.mount.slew_to_target()
+        self.status()  # Seems to help with reading coords
+
+        # Loop over cameras...
+        camera_events = {}
+        for cam_name, exptime in exptimes.items():
+            cam = self.cameras[cam_name]
+
+            # Create filename
+            path = os.path.join(observation.directory, cam.uid, observation.seq_time)
+            filename = os.path.join(
+                path, f'{imtype}_{observation.current_exp_num:02d}.{cam.file_extension}')
+
+            # Take exposure and get event
+            exptime = exptime.to(u.second).value
```

Comment:
I'll go ahead and make the [`to_value`](https://docs.astropy.org/en/stable/api/astropy.units.Quantity.html#astropy.units.Quantity.to_value) suggestion so @AnthonyHorton doesn't have an aneurysm mentioning it again. :stuck_out_tongue_winking_eye: 

```suggestion
            exptime = exptime.to_value(u.second)
```